coc/ext/fullwarapi/__init__.py

```diff
@@ -0,0 +1,221 @@
+    If unsure as to what this means, please reach out to an admin.
+
+    Parameters
+    -----------
+    username : str
+        Your username as given on the discord server.
+    password : str
+        Your password as given on the discord server
+    clash_client: coc.Client
+        Client to use
+
+    loop : Optional[:class:`asyncio.AbstractEventLoop`]
+        The :class:`asyncio.AbstractEventLoop` to use for HTTP requests.
+        An :func:`asyncio.get_event_loop()` will be used if ``None`` is passed
+
+    """
+
+    BASE_URL = "https://fw-api.teamutils.com"
+
+    __slots__ = ("username", "password", "clash_client", "loop", "key", "http_session")
+
+    def __init__(self, username: str, password: str, clash_client, loop: asyncio.AbstractEventLoop = None):
+        self.username = username
+        self.password = password
+        self.clash_client = clash_client
+
+        self.loop = loop or asyncio.get_event_loop()
+        self.key = None  # set in get_key()
+
+        self.http_session = aiohttp.ClientSession(loop=self.loop)
+
+    async def close(self):
+        """Close the client session established"""
+        await self.http_session.close()
+
+    async def _request(self, method, url, *, token_request: bool = False, **kwargs):
+        url = self.BASE_URL + url
+
+        if not token_request:
+            key = await self._get_key()
+
+            headers = {"authorization": "Bearer {}".format(key)}
+            kwargs["headers"] = headers
+
+        async with self.http_session.request(method, url, **kwargs) as response:
+            LOG.debug("%s (%s) has returned %s", url, method, response.status)
+            data = await json_or_text(response)
+            LOG.debug(data)
+
+            if 200 <= response.status < 300:
+                LOG.debug("%s has received %s", url, data)
+                return data
+
+            if response.status == 401:
+                await self._refresh_key()
+                return await self._request(method, url, **kwargs)
+
+    async def _get_key(self):
+        if not self.key or (self.key.expires_at < datetime.utcnow()):
+            await self._refresh_key()
+
+        return self.key.token
+
+    async def _refresh_key(self):
+        data = {
+            "username": self.username,
+            "password": self.password,
+        }
+
+        payload = await self._request("POST", "/login", token_request=True, json=data)
+        self.key = AccessToken(payload["access_token"], extract_expiry_from_jwt_token(payload["access_token"]))
+
+    async def war_result(self, clan_tag: str, preparation_start: int = 0) -> ClanWar:
+        """Get a stored war result.
+
+        Parameters
+        ----------
+        client: coc.Client
+            instance of the clash client
+        clan_tag: str
+            The clan tag to find war result for.
+        preparation_start: int
+            Preparation start of a specific war result to find.
```

Comment:
Is this meant to be a timestamp integer? If so I feel like it makes sense to pass in a datetime object and then have coc.py do the converting into an integer, since it parses the timestamps as datetime objects when getting data from the API. Either way, clarifying what this type is meant to be would be good